torch/csrc/profiler/python/init.cpp

```diff
@@ -10,6 +10,105 @@
 #include <torch/csrc/profiler/standalone/execution_trace_observer.h>
 #include <torch/csrc/utils/pybind.h>
 
+struct THPCapturedTraceback {
+  PyObject_HEAD std::shared_ptr<torch::CapturedTraceback> data;
+};
+
+static int THPCapturedTraceback_traverse(
+    PyObject* self,
+    visitproc visit,
+    void* arg) {
+  return ((THPCapturedTraceback*)self)
+      ->data->traversePython((int (*)(void*, void*))visit, arg);
+}
+
+static int THPCapturedTraceback_clear(PyObject* self) {
+  return ((THPCapturedTraceback*)self)->data->clearPython();
+}
+
+static void THPCapturedTraceback_dealloc(PyObject* self_) {
+  auto* self = (THPCapturedTraceback*)self_;
+  PyObject_GC_UnTrack(self);
+  self->data.~shared_ptr<torch::CapturedTraceback>();
+  // promptly trigger delayed frees since we have GIL
+  torch::freeDeadCapturedTracebackFrames();
+  PyObject_GC_Del(self);
+}
+
+PyTypeObject THPCapturedTracebackType = {
+    PyVarObject_HEAD_INIT(
+        NULL,
+        0) "torch._C._profiler.CapturedTraceback", /* tp_name */
+    sizeof(THPCapturedTraceback), /* tp_basicsize */
+    0, /* tp_itemsize */
+    THPCapturedTraceback_dealloc, /* tp_dealloc */
+    0, /* tp_vectorcall_offset */
+    nullptr, /* tp_getattr */
+    nullptr, /* tp_setattr */
+    nullptr, /* tp_reserved */
+    nullptr, /* tp_repr */
+    nullptr, /* tp_as_number */
+    nullptr, /* tp_as_sequence */
+    nullptr, /* tp_as_mapping */
+    nullptr, /* tp_hash  */
+    nullptr, /* tp_call */
+    nullptr, /* tp_str */
+    nullptr, /* tp_getattro */
+    nullptr, /* tp_setattro */
+    nullptr, /* tp_as_buffer */
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC, /* tp_flags */
+    nullptr, /* tp_doc */
+    (traverseproc)THPCapturedTraceback_traverse, /* tp_traverse */
+    (inquiry)THPCapturedTraceback_clear, /* tp_clear */
+    nullptr, /* tp_richcompare */
+    0, /* tp_weaklistoffset */
+    nullptr, /* tp_iter */
+    nullptr, /* tp_iternext */
+    nullptr, /* tp_methods */
+    nullptr, /* tp_members */
+    nullptr, /* tp_getset */
+    nullptr, /* tp_base */
+    nullptr, /* tp_dict */
+    nullptr, /* tp_descr_get */
+    nullptr, /* tp_descr_set */
+    0, /* tp_dictoffset */
+    nullptr, /* tp_init */
+    nullptr, /* tp_alloc */
+    nullptr, /* tp_new */
+};
+
+namespace pybind11 {
+namespace detail {
+
+template <>
+struct type_caster<std::shared_ptr<torch::CapturedTraceback>> {
```

Comment:
There is a story that all the caster must be in https://github.com/pytorch/pytorch/blob/main/torch/csrc/utils/pybind.h IIRC ?

Reply:
In this case it's OK because these are the only sites we actually use CapturedTraceback in pybind11 (it is also not so convenient to move). If we end up having more CapturedTraceback bindings it can move.